libraries/tickfs/src/flash_controller.rs

```diff
@@ -0,0 +1,79 @@
+//! The Flash Controller interface with hardware
+
+use crate::error_codes::ErrorCode;
+
+/// Implementation required for the flash controller hardware. This
+/// should read, write and erase flash from the hardware using the
+/// flash controller.
+///
+/// This is the public trait for the Flash controller implementation.
+///
+/// The PageSize 'P' must be the smallest size that can be erased in
+/// a single operation.
+/// The start and end address of the FlashController must be aligned
+/// to the PageSize.
+/// All `region_number`s and `address`es are offset from zero. If you
+/// want to use flash that doesn't start at zero, or is a partition
+/// offset from the start of flash you will need to add that offset
+/// to the values in your implementation.
+///
+/// The boiler plate for an implementation will look something like this
+///
+/// ```rust
+/// use tickfs::error_codes::ErrorCode;
+/// use tickfs::flash_controller::FlashController;
+///
+/// #[derive(Default)]
+/// struct FlashCtrl {}
+///
+/// impl FlashCtrl {
+///     fn new() -> Self {
+///         Self { /* fields */ }
+///     }
+/// }
+///
+/// impl FlashController for FlashCtrl {
+///     fn read_region(&self, region_number: usize, offset: usize, buf: &mut [u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+///
+///     fn erase_region(&self, region_number: usize) -> Result<(), ErrorCode> {
+///         unimplemented!()
+///     }
+/// }
+/// ```
+pub trait FlashController {
+    /// This function must read the data from the flash region specified by
+    /// `region_number` into `buf`. The length of the data read should be the
+    /// same length as buf. `offset` indicates an offset into the region that
+    /// should be read.
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::ReadFail.
+    fn read_region(
+        &self,
+        region_number: usize,
+        offset: usize,
+        buf: &mut [u8],
+    ) -> Result<(), ErrorCode>;
+
+    /// This function must write the length of `buf` to the specified address
+    /// in flash.
+    /// If the length of `buf` is smaller then the minimum supported write size
+    /// the implementation can write a larger value. This should be done by first
+    /// reading the value, making the changed from `buf` and then writing it back.
+    ///
+    /// On success it should return nothing, on failure it
+    /// should return ErrorCode::WriteFail.
+    fn write_region(&self, address: usize, buf: &[u8]) -> Result<(), ErrorCode>;
```

Comment:
This doesn't need a region num?

Reply:
No, it's much more efficient to just write the smallest amount of data possible to a specific address. Generally flash controllers don't require an entire page write.

Reply:
Maybe this should just be called `write()` then?

Reply:
Good point, I'll fix this.